### PR TITLE
server: migrate some accesses of randomgenerator to the api interface

### PR DESCRIPTION
**Pull request by akonradi**
Commit Message: Migrate some accesses of RandomGenerator to use the Api interface method
Additional Description:
Continue on #13243 by removing RandomGenerator from some interfaces and changing
call sites to call Api::randomGenerator() instead.

Risk Level: low - there is only ever a single real RandomGenerator
Testing: ran affected tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

File: source/common/upstream/health_discovery_service.cc [hunk 1]
```diff
@@ -273,18 +272,18 @@ HdsCluster::HdsCluster(Server::Admin& admin, Runtime::Loader& runtime,
...
       initial_hosts_(new HostVector()), validation_visitor_(validation_visitor) {
   ENVOY_LOG(debug, "Creating an HdsCluster");
   priority_set_.getOrCreateHostSet(0);
 
-  info_ = info_factory.createClusterInfo(
-      {admin, runtime_, cluster_, bind_config_, stats_, ssl_context_manager_, added_via_api_, cm,
-       local_info, dispatcher, random, singleton_manager, tls, validation_visitor, api});
+  info_ = info_factory.createClusterInfo({admin, runtime_, cluster_, bind_config_, stats_,
+                                          ssl_context_manager_, added_via_api_, cm, local_info,
+                                          dispatcher, api.randomGenerator(), singleton_manager, tls,
```
Comment (jmarantz) [inline source/common/upstream/health_discovery_service.cc:285]: you are passing api below, so maybe you don't need to pass the RNG separately?

File: test/test_common/utility.h [hunk 2]
```diff
@@ -1040,7 +1040,9 @@ makeHeaderMap(const std::initializer_list<std::pair<std::string, std::string>>&
 
 namespace Api {
 ApiPtr createApiForTest();
+ApiPtr createApiForTest(Random::RandomGenerator& random);
 ApiPtr createApiForTest(Stats::Store& stat_store);
+ApiPtr createApiForTest(Stats::Store& stat_store, Random::RandomGenerator& random);
```
Comment (jmarantz) [inline test/test_common/utility.h:1045]: this is only continuing a pattern that can be blamed on me, but I wonder if we should now convert to cxx17 call-style here, e.g. https://stackoverflow.com/questions/6181715/convenient-c-struct-initialisation

File: test/extensions/clusters/aggregate/cluster_update_test.cc [hunk 3]
```diff
@@ -170,22 +171,22 @@ TEST_F(AggregateClusterUpdateTest, LoadBalancingTest) {
 
   Upstream::HostConstSharedPtr host;
   for (int i = 0; i < 33; ++i) {
-    EXPECT_CALL(factory_.random_, random()).WillRepeatedly(Return(i));
+    EXPECT_CALL(random_, random()).WillRepeatedly(Return(i));
```
Comment (jmarantz) [inline test/extensions/clusters/aggregate/cluster_update_test.cc:174]: it is not obvious to me why all these call-sites in this file and a similar one below had to change.

**Review by jmarantz — COMMENTED** [review]
basically looks great with a nit and a small suggestion.

File: source/common/upstream/health_discovery_service.cc (hunk 1 above)
Comment (akonradi) [reply source/common/upstream/health_discovery_service.cc:285]: I'm splitting this refactor across a few PRs, gonna address this one in the next PR.

Comment (jmarantz) [reply source/common/upstream/health_discovery_service.cc:285]: The suggestion here is to make *less* changes in this PR, right?

File: test/test_common/utility.h (hunk 2 above)
Comment (akonradi) [reply test/test_common/utility.h:1045]: Does Envoy style allow designated initializers? They're part of the C++20 spec and the Google style guide has a special exemption for them.

File: test/extensions/clusters/aggregate/cluster_update_test.cc (hunk 3 above)
Comment (akonradi) [reply test/extensions/clusters/aggregate/cluster_update_test.cc]: They didn't, I just split my large refactor in the wrong direction.

File: source/common/upstream/health_discovery_service.cc (hunk 1 above)
Comment (jmarantz) [reply source/common/upstream/health_discovery_service.cc]: oh, never mind. I mis-read the 'before' side. What you have is fine for now given there is a follow-up planned.

Comment (akonradi) [reply source/common/upstream/health_discovery_service.cc]: Fixed, I didn't realize the params object could be modified without changing a bunch of call sites.

**Review by jmarantz — APPROVED** [review]
@envoyproxy/senior-maintainers

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**